// igniteui-angular/esm5/lib/grids/resizing/resizer.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ViewChild } from '@angular/core';
import { IgxColumnResizingService } from './resizing.service';
import { IgxColumnResizerDirective } from './resizer.directive';
var IgxGridColumnResizerComponent = /** @class */ (function () {
    function IgxGridColumnResizerComponent(colResizingService) {
        this.colResizingService = colResizingService;
    }
    IgxGridColumnResizerComponent.ctorParameters = function () { return [
        { type: IgxColumnResizingService }
    ]; };
    __decorate([
        ViewChild(IgxColumnResizerDirective, { static: true })
    ], IgxGridColumnResizerComponent.prototype, "resizer", void 0);
    IgxGridColumnResizerComponent = __decorate([
        Component({
            changeDetection: ChangeDetectionStrategy.OnPush,
            preserveWhitespaces: false,
            selector: 'igx-grid-column-resizer',
            template: "<div class=\"igx-grid__th-resize-line\"\n    [style.left.px]=\"-99999\" igxResizer\n    [style.height.px]=\"colResizingService.resizerHeight\"\n    [restrictHResizeMax]=\"colResizingService.restrictResizeMax\"\n    [restrictHResizeMin]=\"colResizingService.restrictResizeMin\"\n    (resizeEnd)=\"colResizingService.resizeColumn($event)\">\n</div>\n"
        })
    ], IgxGridColumnResizerComponent);
    return IgxGridColumnResizerComponent;
}());
export { IgxGridColumnResizerComponent };
