// igniteui-angular/esm5/lib/grids/resizing/resize-handle.directive.js
import { __decorate } from "tslib";
import { AfterViewInit, Directive, ElementRef, Input, NgZone, HostListener, OnDestroy } from '@angular/core';
import { Subject, fromEvent } from 'rxjs';
import { debounceTime, takeUntil } from 'rxjs/operators';
import { IgxColumnResizingService } from './resizing.service';
/**
 * @hidden
 * @internal
 */
var IgxResizeHandleDirective = /** @class */ (function () {
    function IgxResizeHandleDirective(zone, element, colResizingService) {
        this.zone = zone;
        this.element = element;
        this.colResizingService = colResizingService;
        /**
         * @hidden
         */
        this._dblClick = false;
        /**
         * @hidden
         */
        this.destroy$ = new Subject();
        this.DEBOUNCE_TIME = 200;
    }
    /**
     * @hidden
     */
    IgxResizeHandleDirective.prototype.ngOnDestroy = function () {
        this.destroy$.next(true);
        this.destroy$.complete();
    };
    /**
     * @hidden
     */
    IgxResizeHandleDirective.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (!this.column.columnGroup && this.column.resizable) {
            this.zone.runOutsideAngular(function () {
                fromEvent(_this.element.nativeElement, 'mousedown').pipe(debounceTime(_this.DEBOUNCE_TIME), takeUntil(_this.destroy$)).subscribe(function (event) {
                    if (_this._dblClick) {
                        _this._dblClick = false;
                        return;
                    }
                    if (event.button === 0) {
                        _this._onResizeAreaMouseDown(event);
                        _this.column.grid.resizeLine.resizer.onMousedown(event);
                    }
                });
            });
            fromEvent(this.element.nativeElement, 'mouseup').pipe(debounceTime(this.DEBOUNCE_TIME), takeUntil(this.destroy$)).subscribe(function () {
                _this.colResizingService.isColumnResizing = false;
                _this.colResizingService.showResizer = false;
                _this.column.grid.cdr.detectChanges();
            });
        }
    };
    /**
     * @hidden
     */
    IgxResizeHandleDirective.prototype.onMouseOver = function () {
        this.colResizingService.resizeCursor = 'col-resize';
    };
    /**
     * @hidden
     */
    IgxResizeHandleDirective.prototype.onDoubleClick = function () {
        this._dblClick = true;
        this.colResizingService.column = this.column;
        this.colResizingService.autosizeColumnOnDblClick();
    };
    /**
     * @hidden
     */
    IgxResizeHandleDirective.prototype._onResizeAreaMouseDown = function (event) {
        this.colResizingService.column = this.column;
        this.colResizingService.isColumnResizing = true;
        this.colResizingService.startResizePos = event.clientX;
        this.colResizingService.showResizer = true;
        this.column.grid.cdr.detectChanges();
    };
    IgxResizeHandleDirective.ctorParameters = function () { return [
        { type: NgZone },
        { type: ElementRef },
        { type: IgxColumnResizingService }
    ]; };
    __decorate([
        Input('igxResizeHandle')
    ], IgxResizeHandleDirective.prototype, "column", void 0);
    __decorate([
        HostListener('mouseover')
    ], IgxResizeHandleDirective.prototype, "onMouseOver", null);
    __decorate([
        HostListener('dblclick')
    ], IgxResizeHandleDirective.prototype, "onDoubleClick", null);
    IgxResizeHandleDirective = __decorate([
        Directive({
            selector: '[igxResizeHandle]'
        })
    ], IgxResizeHandleDirective);
    return IgxResizeHandleDirective;
}());
export { IgxResizeHandleDirective };
