// igniteui-angular/esm5/lib/grids/column-chooser-base.js
import { __decorate, __extends, __values } from "tslib";
import { ChangeDetectorRef, HostBinding, Input, OnDestroy, Directive } from '@angular/core';
import { IgxStringFilteringOperand } from '../data-operations/filtering-condition';
import { FilteringExpressionsTree } from '../data-operations/filtering-expressions-tree';
import { FilteringLogic } from '../data-operations/filtering-expression.interface';
import { FilteringStrategy } from '../data-operations/filtering-strategy';
import { ColumnDisplayOrder } from './common/enums';
var CustomFilteringStrategy = /** @class */ (function (_super) {
    __extends(CustomFilteringStrategy, _super);
    function CustomFilteringStrategy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CustomFilteringStrategy.prototype.filter = function (data, expressionsTree) {
        var _this = this;
        var res = [];
        data.forEach(function (item) {
            if (_this.matchRecord(item, expressionsTree.filteringOperands[0])) {
                res.push(item);
            }
            else if (item.column.columnGroup) {
                if (item.column.allChildren.findIndex(function (child) {
                    return _this.matchRecord(child, expressionsTree.filteringOperands[1]) ||
                        _this.matchRecord(child, expressionsTree.filteringOperands[2]);
                }) > -1) {
                    res.push(item);
                }
            }
        });
        return res;
    };
    return CustomFilteringStrategy;
}(FilteringStrategy));
/** @hidden */
var ColumnChooserBaseDirective = /** @class */ (function () {
    function ColumnChooserBaseDirective(cdr) {
        this.cdr = cdr;
        /**
         * Hides/ shows the filtering columns input from the UI.
         */
        this.disableFilter = false;
        /**
         * Access to the columnHidingUI:
         * ```typescript
         * @ViewChild('column-hiding-component')
         *  public columnHidingUI: IgxColumnHidingComponent;
         * ```
         * Sets/gets the max height of the column area.
         * ```typescript
         * let columnsAreaMaxHeight =  this.columnHidingUI.columnsAreaMaxHeight;
         * ```
         *
         * ```html
         * <igx-column-hiding [columnsAreaMaxHeight]="200px"></igx-column-hiding>
         * ```
         */
        this.columnsAreaMaxHeight = '100%';
        /**
         * Sets/Gets the css class selector.
         * By default the value of the `class` attribute is `"igx-column-hiding"`.
         * ```typescript
         * let cssCLass =  this.columnHidingUI.cssClass;
         * ```
         * ```typescript
         * this.columnHidingUI.cssClass = 'column-chooser';
         * ```
         */
        this.cssClass = 'igx-column-hiding';
        /**
         *@hidden
         */
        this._currentColumns = [];
        /**
         *@hidden
         */
        this._gridColumns = [];
        /**
         *@hidden
         */
        this._rawColumns = [];
        /**
         *@hidden
         */
        this._columnDisplayOrder = ColumnDisplayOrder.DisplayOrder;
        /**
         *@hidden
         */
        this._filterCriteria = '';
        /**
         *@hidden
         */
        this._filterColumnsPrompt = '';
        /**
         *@hidden
         */
        this._title = '';
    }
    Object.defineProperty(ColumnChooserBaseDirective.prototype, "columns", {
        /**
         * Gets the grid columns that are going to be manipulated.
         * ```typescript
         * let gridColumns = this.columnHidingUI.columns;
         * ```
         */
        get: function () {
            return this._gridColumns;
        },
        /**
         * Sets the the grid columns that are going to be manipulated.
         * ```html
         * <igx-column-hiding [columns]="grid.columns"></igx-column-hiding>
         * ```
         */
        set: function (value) {
            if (value) {
                this._gridColumns = value;
                this.createColumnItems();
                if (this.filterCriteria) {
                    this.filter();
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnChooserBaseDirective.prototype, "title", {
        /**
         * Sets/gets the title of the column chooser.
         * ```typescript
         * let title =  this.columnHidingUI.title;
         * ```
         */
        get: function () {
            return this._title;
        },
        /**
         * ```html
         * <igx-column-hiding [title]="'IgxColumnHidingComponent Title'"></igx-column-hiding>
         * ```
         */
        set: function (value) {
            this._title = (value) ? value : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnChooserBaseDirective.prototype, "filterColumnsPrompt", {
        /**
         * Gets the prompt that is displayed in the filter input.
         * ```typescript
         * let filterColumnsPrompt =  this.columnHidingUI.filterColumnsPrompt;
         * ```
         */
        get: function () {
            return this._filterColumnsPrompt;
        },
        /**
         * Sets the prompt that is going to be displayed in the filter input.
         * ```html
         * <igx-column-hiding [filterColumnsPrompt]="'Type here to search'"></igx-column-hiding>
         * ```
         */
        set: function (value) {
            this._filterColumnsPrompt = (value) ? value : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnChooserBaseDirective.prototype, "columnItems", {
        /**
         * Gets the items of the selected columns.
         * ```typescript
         * let columnItems =  this.columnHidingUI.columnItems;
         * ```
         */
        get: function () {
            return this._currentColumns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnChooserBaseDirective.prototype, "filterCriteria", {
        /**
         * Gets the value which filters the columns list.
         * ```typescript
         * let filterCriteria =  this.columnHidingUI.filterCriteria;
         * ```
         */
        get: function () {
            return this._filterCriteria;
        },
        /**
         * Sets the value which filters the columns list.
         * ```html
         *  <igx-column-hiding [filterCriteria]="'ID'"></igx-column-hiding>
         * ```
         */
        set: function (value) {
            if (!value || value.length === 0) {
                this.clearFiltering();
                this._filterCriteria = '';
                this.cdr.detectChanges();
                return;
            }
            else if (this._filterCriteria && this._filterCriteria.length > value.length) {
                this.clearFiltering();
            }
            this._filterCriteria = value;
            this.filter();
            this.cdr.detectChanges();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnChooserBaseDirective.prototype, "columnDisplayOrder", {
        /**
         * Gets the display order of the columns.
         * ```typescript
         * let columnDisplayOrder  =  this.columnHidingUI.columnDisplayOrder;
         * ```
         */
        get: function () {
            return this._columnDisplayOrder;
        },
        /**
         * Sets the display order of the columns.
         * ```typescript
         * this.columnHidingUI.columnDisplayOrder = ColumnDisplayOrder.Alphabetical;
         * ```
         */
        set: function (value) {
            if (value !== undefined) {
                this.orderColumns(value);
                if (this._filterCriteria.length > 0) {
                    this.filter();
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     *@hidden
     */
    ColumnChooserBaseDirective.prototype.ngOnDestroy = function () {
        var e_1, _a;
        try {
            for (var _b = __values(this._currentColumns), _c = _b.next(); !_c.done; _c = _b.next()) {
                var item = _c.value;
                item.valueChanged.unsubscribe();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     *@hidden
     */
    ColumnChooserBaseDirective.prototype.createColumnItems = function () {
        var _this = this;
        if (this._gridColumns.length > 0) {
            this._rawColumns = [];
            this._gridColumns.forEach(function (column) {
                var item = _this.createColumnItem(_this, column);
                if (item) {
                    _this._rawColumns.push(item);
                }
            });
            this._currentColumns = this._rawColumns.slice(0);
            this.orderColumns(this._columnDisplayOrder);
        }
    };
    /**
     *@hidden
     */
    ColumnChooserBaseDirective.prototype.orderColumns = function (value) {
        this._columnDisplayOrder = value;
        if (value === ColumnDisplayOrder[ColumnDisplayOrder.Alphabetical] ||
            value === ColumnDisplayOrder.Alphabetical) {
            this._currentColumns = this._rawColumns.slice(0).sort(function (current, next) {
                return current.name.toLowerCase().localeCompare(next.name.toLowerCase());
            });
        }
        else {
            this._currentColumns = this._rawColumns;
        }
    };
    /**
     *@hidden
     */
    ColumnChooserBaseDirective.prototype.filter = function () {
        var filteringExpressionsTree = new FilteringExpressionsTree(FilteringLogic.Or);
        filteringExpressionsTree.filteringOperands.push(this.createFilteringExpression('name'));
        filteringExpressionsTree.filteringOperands.push(this.createFilteringExpression('field'));
        filteringExpressionsTree.filteringOperands.push(this.createFilteringExpression('header'));
        var strategy = new CustomFilteringStrategy();
        this._currentColumns = strategy.filter(this._currentColumns, filteringExpressionsTree);
    };
    /**
     *@hidden
     */
    ColumnChooserBaseDirective.prototype.createFilteringExpression = function (fieldName) {
        return {
            condition: IgxStringFilteringOperand.instance().condition('contains'),
            fieldName: fieldName,
            ignoreCase: true,
            searchVal: this._filterCriteria
        };
    };
    /**
     *@hidden
     */
    ColumnChooserBaseDirective.prototype.clearFiltering = function () {
        this.createColumnItems();
    };
    ColumnChooserBaseDirective.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input()
    ], ColumnChooserBaseDirective.prototype, "columns", null);
    __decorate([
        Input()
    ], ColumnChooserBaseDirective.prototype, "title", null);
    __decorate([
        Input()
    ], ColumnChooserBaseDirective.prototype, "filterColumnsPrompt", null);
    __decorate([
        Input()
    ], ColumnChooserBaseDirective.prototype, "disableFilter", void 0);
    __decorate([
        Input()
    ], ColumnChooserBaseDirective.prototype, "columnItems", null);
    __decorate([
        Input()
    ], ColumnChooserBaseDirective.prototype, "filterCriteria", null);
    __decorate([
        Input()
    ], ColumnChooserBaseDirective.prototype, "columnDisplayOrder", null);
    __decorate([
        Input()
    ], ColumnChooserBaseDirective.prototype, "columnsAreaMaxHeight", void 0);
    __decorate([
        HostBinding('attr.class')
    ], ColumnChooserBaseDirective.prototype, "cssClass", void 0);
    ColumnChooserBaseDirective = __decorate([
        Directive()
    ], ColumnChooserBaseDirective);
    return ColumnChooserBaseDirective;
}());
export { ColumnChooserBaseDirective };
