// igniteui-angular/esm5/lib/services/direction/directionality.js
import { __decorate, __param } from "tslib";
import { Injectable, Inject, InjectionToken, inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import * as i0 from "@angular/core";
/**
 * Injection token is used to inject the document into Directionality
 * which factory could be faked for testing purposes.
 *
 * We can't provide and mock the DOCUMENT token from platform-browser because configureTestingModule
 * allows override of the default providers, directive, pipes, modules of the test injector
 * which causes errors.
 *
 * @hidden
 */
export var DIR_DOCUMENT = new InjectionToken('dir-doc', {
    providedIn: 'root',
    factory: DIR_DOCUMENT_FACTORY
});
/**
 * @hidden
 */
export function DIR_DOCUMENT_FACTORY() {
    return inject(DOCUMENT);
}
/**
 * @hidden
 *
 * Bidirectional service that extracts the value of the direction attribute on the body or html elements.
 *
 * The dir attribute over the body element takes precedence.
 */
var IgxDirectionality = /** @class */ (function () {
    function IgxDirectionality(document) {
        this._document = document;
        var bodyDir = this._document.body ? this._document.body.dir : null;
        var htmlDir = this._document.documentElement ? this._document.documentElement.dir : null;
        var extractedDir = bodyDir || htmlDir;
        this._dir = (extractedDir === 'ltr' || extractedDir === 'rtl') ? extractedDir : 'ltr';
    }
    Object.defineProperty(IgxDirectionality.prototype, "value", {
        get: function () {
            return this._dir;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDirectionality.prototype, "document", {
        get: function () {
            return this._document;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDirectionality.prototype, "rtl", {
        get: function () {
            return this._dir === 'rtl';
        },
        enumerable: true,
        configurable: true
    });
    IgxDirectionality.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [DIR_DOCUMENT,] }] }
    ]; };
    IgxDirectionality.ɵprov = i0.ɵɵdefineInjectable({ factory: function IgxDirectionality_Factory() { return new IgxDirectionality(i0.ɵɵinject(DIR_DOCUMENT)); }, token: IgxDirectionality, providedIn: "root" });
    IgxDirectionality = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __param(0, Inject(DIR_DOCUMENT))
    ], IgxDirectionality);
    return IgxDirectionality;
}());
export { IgxDirectionality };
