// igniteui-angular/esm5/lib/tabs/tabs.common.js
/** @hidden */
var IgxTabsBase = /** @class */ (function () {
    function IgxTabsBase() {
    }
    IgxTabsBase.prototype.scrollElement = function (element, scrollRight) { };
    IgxTabsBase.prototype.performSelectionChange = function (newTab) { };
    IgxTabsBase.prototype.transformContentAnimation = function (tab, duration) { };
    IgxTabsBase.prototype.transformIndicatorAnimation = function (element, duration) { };
    return IgxTabsBase;
}());
export { IgxTabsBase };
/** @hidden */
var IgxTabItemBase = /** @class */ (function () {
    function IgxTabItemBase() {
    }
    Object.defineProperty(IgxTabItemBase.prototype, "index", {
        get: function () { return 0; },
        enumerable: true,
        configurable: true
    });
    IgxTabItemBase.prototype.select = function () { };
    IgxTabItemBase.prototype.setSelectedInternal = function (newValue) { };
    return IgxTabItemBase;
}());
export { IgxTabItemBase };
/** @hidden */
var IgxTabsGroupBase = /** @class */ (function () {
    function IgxTabsGroupBase() {
    }
    IgxTabsGroupBase.prototype.select = function () { };
    IgxTabsGroupBase.prototype.setSelectedInternal = function (newValue) { };
    return IgxTabsGroupBase;
}());
export { IgxTabsGroupBase };
