// igniteui-angular/esm5/lib/services/exporter-common/exporter-options-base.js
var IgxExporterOptionsBase = /** @class */ (function () {
    function IgxExporterOptionsBase(fileName, _fileExtension) {
        this._fileExtension = _fileExtension;
        /**
         * Specifies whether hidden columns should be exported.
         * ```typescript
         * let ignoreColumnsVisibility = this.exportOptions.ignoreColumnsVisibility;
         * this.exportOptions.ignoreColumnsVisibility = true;
         * ```
         * @memberof IgxExporterOptionsBase
         */
        this.ignoreColumnsVisibility = false;
        /**
         * Specifies whether filtered out rows should be exported.
         * ```typescript
         * let ignoreFiltering = this.exportOptions.ignoreFiltering;
         * this.exportOptions.ignoreFiltering = true;
         * ```
         * @memberof IgxExporterOptionsBase
         */
        this.ignoreFiltering = false;
        /**
         * Specifies if the exporter should ignore the current column order in the IgxGrid.
         * ```typescript
         * let ignoreColumnsOrder = this.exportOptions.ignoreColumnsOrder;
         * this.exportOptions.ignoreColumnsOrder = true;
         * ```
         * @memberof IgxExporterOptionsBase
         */
        this.ignoreColumnsOrder = false;
        /**
         * Specifies whether the exported data should be sorted as in the provided IgxGrid.
         * ```typescript
         * let ignoreSorting = this.exportOptions.ignoreSorting;
         * this.exportOptions.ignoreSorting = true;
         * ```
         * @memberof IgxExporterOptionsBase
         */
        this.ignoreSorting = false;
        this.setFileName(fileName);
    }
    IgxExporterOptionsBase.prototype.setFileName = function (fileName) {
        this._fileName = fileName + (fileName.endsWith(this._fileExtension) === false ? this._fileExtension : '');
    };
    Object.defineProperty(IgxExporterOptionsBase.prototype, "fileName", {
        /**
         * Gets the file name which will be used for the exporting operation.
         * ```typescript
         * let fileName = this.exportOptions.fileName;
         * ```
         * @memberof IgxExporterOptionsBase
         */
        get: function () {
            return this._fileName;
        },
        /**
         * Sets the file name which will be used for the exporting operation.
         * ```typescript
         * this.exportOptions.fileName = 'exportedData01';
         * ```
         * @memberof IgxExporterOptionsBase
         */
        set: function (value) {
            this.setFileName(value);
        },
        enumerable: true,
        configurable: true
    });
    return IgxExporterOptionsBase;
}());
export { IgxExporterOptionsBase };
