// igniteui-angular/esm2015/lib/services/excel/excel-folders.js
import { ExcelFileTypes, ExcelFolderTypes } from './excel-enums';
/** @hidden */
export class RootExcelFolder {
    get folderName() {
        return '';
    }
    childFiles(data) {
        return [ExcelFileTypes.ContentTypesFile];
    }
    childFolders(data) {
        return [
            ExcelFolderTypes.RootRelsExcelFolder,
            ExcelFolderTypes.DocPropsExcelFolder,
            ExcelFolderTypes.XLExcelFolder
        ];
    }
}
/** @hidden */
export class RootRelsExcelFolder {
    get folderName() {
        return '_rels';
    }
    childFiles(data) {
        return [ExcelFileTypes.RootRelsFile];
    }
    childFolders(data) {
        return [];
    }
}
/** @hidden */
export class DocPropsExcelFolder {
    get folderName() {
        return 'docProps';
    }
    childFiles(data) {
        return [
            ExcelFileTypes.AppFile,
            ExcelFileTypes.CoreFile
        ];
    }
    childFolders(data) {
        return [];
    }
}
/** @hidden */
export class XLExcelFolder {
    get folderName() {
        return 'xl';
    }
    childFiles(data) {
        const retVal = [
            ExcelFileTypes.StyleFile,
            ExcelFileTypes.WorkbookFile
        ];
        if (!data.isEmpty) {
            retVal.push(ExcelFileTypes.SharedStringsFile);
        }
        return retVal;
    }
    childFolders(data) {
        const retVal = [
            ExcelFolderTypes.XLRelsExcelFolder,
            ExcelFolderTypes.ThemeExcelFolder,
            ExcelFolderTypes.WorksheetsExcelFolder
        ];
        if (!data.isEmpty && data.options.exportAsTable) {
            retVal.push(ExcelFolderTypes.TablesExcelFolder);
        }
        return retVal;
    }
}
/** @hidden */
export class XLRelsExcelFolder {
    get folderName() {
        return '_rels';
    }
    childFiles(data) {
        return [ExcelFileTypes.WorkbookRelsFile];
    }
    childFolders(data) {
        return [];
    }
}
/** @hidden */
export class ThemeExcelFolder {
    get folderName() {
        return 'theme';
    }
    childFiles(data) {
        return [ExcelFileTypes.ThemeFile];
    }
    childFolders(data) {
        return [];
    }
}
/** @hidden */
export class WorksheetsExcelFolder {
    get folderName() {
        return 'worksheets';
    }
    childFiles(data) {
        return [ExcelFileTypes.WorksheetFile];
    }
    childFolders(data) {
        return data.isEmpty || !data.options.exportAsTable ? [] : [ExcelFolderTypes.WorksheetsRelsExcelFolder];
    }
}
/** @hidden */
export class TablesExcelFolder {
    get folderName() {
        return 'tables';
    }
    childFiles(data) {
        return [ExcelFileTypes.TablesFile];
    }
    childFolders(data) {
        return [];
    }
}
/** @hidden */
export class WorksheetsRelsExcelFolder {
    get folderName() {
        return '_rels';
    }
    childFiles(data) {
        return [ExcelFileTypes.WorksheetRelsFile];
    }
    childFolders(data) {
        return [];
    }
}
